gcp_variant_transforms/transforms/infer_undefined_headers.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2018 Google Inc.  All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A PTransform to infer undefined header fields"""
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+
+from vcf.parser import _Format as Format
+from vcf.parser import _Info as Info
+from vcf.parser import field_counts
+
+from gcp_variant_transforms.beam_io import vcf_header_io
+from gcp_variant_transforms.transforms import merge_headers
+
+
+__all__ = ['InferUndefinedHeaderFields']
+
+
+class VcfFieldType(object):
+  """Constants for VCF field types."""
+  FLAG = 'Flag'
+  STRING = 'String'
+
+
+class _InferUndefinedHeaderFields(beam.DoFn):
+  """Extracts undefined header fields from ``Variant`` record."""
+
+  def _get_field_count(self, field_value):
+    """
+    Args:
+      field_value (list, bool, integer, or string): value for the field
+        returned by pyvcf. E.g. [0.33, 0.66] is a field value for Allele freq
+        field (AF).
+    """
+    if isinstance(field_value, list):
+      return field_counts['.']
+    elif isinstance(field_value, bool):
+      return 0
+    else:
+      return 1
+
+  def _get_field_type(self, field_value):
+    """
+    Args:
+      field_value (list, bool, integer, or string): value for the field
+        returned by pyvcf. E.g. [0.33, 0.66] is a field value for Allele freq
+        field (AF).
+    """
+    if isinstance(field_value, bool):
+      return VcfFieldType.FLAG
+    else:
+      return VcfFieldType.STRING
+
+  def _infer_undefined_info_fields(self, variant, defined_headers):
+    """
```

Comment:
Please add a short description as well, since it's not obvious that it only returns fields not defined in the header (for both info and format).

Reply:
Done!